pl_examples/domain_templates/computer_vision_fine_tuning.py

```diff
@@ -0,0 +1,447 @@
+                   train_bn=self.hparams.train_bn)
+
+    def on_epoch_start(self):
+        """Use `on_epoch_start` to unfreeze layers progressively."""
+        optimizer = self.trainer.optimizers[0]
+        if self.current_epoch == self.hparams.milestones[0]:
+            _unfreeze_and_add_param_group(module=self.feature_extractor[-2:],
+                                          optimizer=optimizer,
+                                          train_bn=self.hparams.train_bn)
+
+        elif self.current_epoch == self.hparams.milestones[1]:
+            _unfreeze_and_add_param_group(module=self.feature_extractor[:-2],
+                                          optimizer=optimizer,
+                                          train_bn=self.hparams.train_bn)
+
+    def training_step(self, batch, batch_idx):
+
+        # 1. Forward pass:
+        x, y = batch
+        y_logits = self.forward(x)
+        y_true = y.view((-1, 1)).type_as(x)
+        y_bin = torch.ge(y_logits, 0)
+
+        # 2. Compute loss & accuracy:
+        train_loss = self.loss(y_true, y_logits)
+        num_correct = torch.eq(y_bin.view(-1), y_true.view(-1)).sum().item()
+
+        # 3. Outputs:
+        tqdm_dict = {'train_loss': train_loss}
+        output = OrderedDict({'loss': train_loss,
+                              'num_correct': num_correct,
+                              'log': tqdm_dict,
+                              'progress_bar': tqdm_dict})
+
+        return output
+
+    def training_epoch_end(self, outputs):
+
+        train_loss_mean = 0.
+        train_acc_mean = 0.
+        for output in outputs:
+
+            train_loss = output['loss']
+            train_acc = output['num_correct']
+            # reduce manually when using dp
+            if self.trainer.use_dp or self.trainer.use_ddp2:
+                train_loss = torch.mean(train_loss)
+            train_loss_mean += train_loss
+            train_acc_mean += train_acc
+
+        train_loss_mean /= len(outputs)
+        train_acc_mean /= (len(outputs) * self.hparams.batch_size)
+        return {'log': {'train_loss': train_loss_mean,
+                        'train_acc': train_acc_mean}}
+
+    def validation_step(self, batch, batch_idx):
+
+        # 1. Forward pass:
+        x, y = batch
+        y_logits = self.forward(x)
+        y_true = y.view((-1, 1)).type_as(x)
+        y_bin = torch.ge(y_logits, 0)
+
+        # 2. Compute loss & accuracy:
+        val_loss = self.loss(y_true, y_logits)
+        num_correct = torch.eq(y_bin.view(-1), y_true.view(-1)).sum().item()
+
+        return {'val_loss': val_loss,
+                'num_correct': num_correct}
+
+    def validation_epoch_end(self, outputs):
+
+        val_loss_mean = 0.
+        val_acc_mean = 0.
+        for output in outputs:
+
+            val_loss = output['val_loss']
+            val_acc = output['num_correct']
+            # reduce manually when using dp
+            if self.trainer.use_dp or self.trainer.use_ddp2:
```

Comment:
can you not just do `torch.stack([output[metric] for output in outputs]).mean() `?
Then you include all cases, dp, ddp, ... without all of this code. same applies to the training_epoch_end.

Reply:
Done!

Reply:
I think now you divide by the length twice in case of val_loss (you have the mean above already)

Reply:
Yes, of course! Thanks! It's fixed.